swarm/storage/pushsync/pusher.go

```diff
@@ -0,0 +1,277 @@
+	})
+	defer deregister()
+
+	chunksInBatch := -1
+	var batchStartTime time.Time
+
+	for {
+		select {
+
+		// retry interval timer triggers starting from new
+		case <-timer.C:
+			metrics.GetOrRegisterCounter("pusher.subscribe-push", nil).Inc(1)
+			// TODO: implement some smart retry strategy relying on sent/synced ratio change
+			// if subscribe was running, stop it
+			if stop != nil {
+				stop()
+			}
+			for _, addr := range synced {
+				// set chunk status to synced, insert to db GC index
+				if err := p.store.Set(context.Background(), chunk.ModeSetSync, addr); err != nil {
+					log.Warn("error setting chunk to synced", "addr", addr, "err", err)
+					continue
+				}
+				delete(p.pushed, addr.Hex())
+			}
+			// reset synced list
+			synced = nil
+
+			// we don't want to record the first iteration
+			if chunksInBatch != -1 {
+				// this measurement is not a timer, but we want a histogram, so it fits the data structure
+				metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.hist", nil).Update(time.Duration(chunksInBatch))
+
+				metrics.GetOrRegisterResettingTimer("pusher.subscribe-push.chunks-in-batch.time", nil).UpdateSince(batchStartTime)
+				metrics.GetOrRegisterCounter("pusher.subscribe-push.chunks-in-batch", nil).Inc(int64(chunksInBatch))
+			}
+			chunksInBatch = 0
+			batchStartTime = time.Now()
+
+			// and start iterating on Push index from the beginning
+			ctx, cancel = context.WithCancel(context.Background())
+			chunks, stop = p.store.SubscribePush(ctx)
+			// reset timer to go off after retryInterval
+			timer.Reset(retryInterval)
+
+		// handle incoming chunks
+		case ch, more := <-chunks:
+			chunksInBatch++
+			// if no more, set to nil and wait for timer
+			if !more {
+				chunks = nil
+				continue
+			}
+
+			metrics.GetOrRegisterCounter("pusher.send-chunk", nil).Inc(1)
+			// if no need to sync this chunk then continue
+			if !p.needToSync(ch) {
+				continue
+			}
+
+			metrics.GetOrRegisterCounter("pusher.send-chunk.send-to-sync", nil).Inc(1)
+			// send the chunk and ignore the error
+			if err := p.sendChunkMsg(ch); err != nil {
+				log.Error("error sending chunk", "addr", ch.Address(), "err", err)
+			}
+
+		// handle incoming receipts
+		case addr := <-p.receipts:
+			metrics.GetOrRegisterCounter("pusher.receipts.all", nil).Inc(1)
+			log.Debug("synced", "addr", addr)
+			// ignore if already received receipt
+			item, found := p.pushed[addr.Hex()]
+			if !found {
+				metrics.GetOrRegisterCounter("pusher.receipts.not-found", nil).Inc(1)
+				log.Debug("not wanted or already got... ignore", "addr", addr)
+				continue
+			}
+			if item.synced {
+				metrics.GetOrRegisterCounter("pusher.receipts.already-synced", nil).Inc(1)
+				log.Debug("just synced... ignore", "addr", addr)
+				continue
+			}
+			metrics.GetOrRegisterCounter("pusher.receipts.synced", nil).Inc(1)
+			// collect synced addresses
+			synced = append(synced, addr)
+			// set synced flag
+			item.synced = true
```

Comment:
See my comment above, and question to Viktor. We seem to be collecting all `synced` chunks so that we can mark the in the LocalStore as synced efficiently in a batch. I am not convinced we need this.